torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -1471,48 +1478,58 @@ def get_original_names_and_sharded_parameters(self):
+    The output from this method will be used in ``consolidate_sharded_model_checkpoints``.
     """
     shard_info = {}
     flatten_info = {}
     buffer_info = {}
-    for module_name, m in self.named_modules():  # includes self
-      # remove "_fpw_module." from module names since it is also removed in
-      # XlaFullyShardedDataParallel's state_dict()
-      module_name = module_name.replace("_fpw_module.", "")
-
-      if isinstance(m, XlaFullyShardedDataParallel):
-        sharded_param_info = {}
-        for p_shard in m.sharded_params:
-          sharded_param_info[p_shard._name] = {
-              "_orig_size": p_shard._orig_size,
-              "_orig_name": p_shard._orig_name,
-          }
-        shard_info[module_name] = sharded_param_info
-
-      if isinstance(m, XlaFlattenParamsWrapper):
-        for i in range(len(m.flat_params)):
-          param_name = f"flat_param_{i}"
-          if module_name != "":
-            param_name = module_name + "." + param_name
-          flatten_info[param_name] = m.metadata(i)
-
-    for name, buf in self.named_buffers():
+    # remove "_fpw_module." from module names since it is also removed in
+    # XlaFullyShardedDataParallel's state_dict()
+    module_name = prefix.replace("_fpw_module.", "")
+    # remove the trailing "." since state_dict will add it
+    module_name = module_name.rstrip(".")
+
+    if isinstance(m, XlaFullyShardedDataParallel):
+      sharded_param_info = {}
+      for p_shard in m.sharded_params:
+        sharded_param_info[p_shard._name] = {
+            "_orig_size": p_shard._orig_size,
+            "_orig_name": p_shard._orig_name,
+        }
+      shard_info[module_name] = sharded_param_info
+
+    if isinstance(m, XlaFlattenParamsWrapper):
+      for i in range(len(m.flat_params)):
+        param_name = f"flat_param_{i}"
+        if module_name != "":
+          param_name = module_name + "." + param_name
+        flatten_info[param_name] = m.metadata(i)
+
+    for name, buf in m.named_buffers(recurse=False):
       if buf is not None and hasattr(buf, "_orig_dtype"):
-        buffer_info[name] = {"_orig_dtype": buf._orig_dtype}
-
-    metadata = {
-        "shard_info": shard_info,
-        "flatten_info": flatten_info,
-        "buffer_info": buffer_info,
-        "world_size": self.world_size,
-        "rank": self.rank,
-    }
-    return metadata
+        buffer_info[prefix + name] = {"_orig_dtype": buf._orig_dtype}
+
+    if "shard_metadata" not in destination:
+      destination["shard_metadata"] = {
+          "shard_info": shard_info,
+          "flatten_info": flatten_info,
+          "buffer_info": buffer_info,
+          "world_size": m.world_size,
+          "rank": m.rank,
+      }
+    else:
+      destination["shard_metadata"]["shard_info"].update(shard_info)
+      destination["shard_metadata"]["flatten_info"].update(flatten_info)
+      destination["shard_metadata"]["buffer_info"].update(buffer_info)
```

Comment:
And would this cause a name conflict if there is a parameter or a buffer under the same named `"shard_metadata"`?